Dewikification: Content app improvements.

kosayoda opened the pull request:
### Simplifications
- Metadata for categories now use `title` as the key to match page metadata, previously it was `name`.
- Rather than having two separate keys for the Font Awesome icon class and icon, it is now under `icon`.

### Structure enhancements
This PR adds a way to have a path be both a category and a page. This is useful for places like `/contributing`, where https://pythondiscord.com/pages/contributing/ is a rendered page while still being a category for https://pythondiscord.com/pages/contributing/sir-lancebot/ and the like.

### Markdown additions
Certain markdown-to-html additions were made to enable feature parity with the current `wiki`-ed site guides, but without the article writer needing to write HTML.

Specifically, these features are added:
- Image captions
![image](https://user-images.githubusercontent.com/41782385/112822413-7ec3a580-90ba-11eb-859b-0ef3a0708ad4.png)
- Table of contents
![image](https://user-images.githubusercontent.com/41782385/112822494-9b5fdd80-90ba-11eb-949b-a1baffbda4ad.png)

kosayoda committed:
Improve content page and listing metadata keys.

Make category `title` consistent with page `title`.
Simplify `icon` key, since there is no benefit to defining the icon
class and the icon separately.
Allow overriding the `icon` for category entries just like page entries.

kosayoda committed:
Allow displaying a page at a category path.

This is useful for our /contributing pages, which have a main
contributing page at /contributing but project pages under the same
path, /contributing/bot etc.

kosayoda committed:
Document additional markdown features.

Adds image caption capabilities and documents missing HTML extensions
available.

kosayoda committed:
Allow adding a table of contents to a page.

kosayoda committed:
Fix failing tests.

kosayoda committed:
Add new tests to achieve full coverage.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -22,17 +22,38 @@ As website changes require staff approval, discussing the page content beforehan
 
 All the markdown files in this folder will then be under this category.
 
+#### Having the Category also be a Page
```
```suggestion
#### Having the Category Also Be a Page
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
````diff
@@ -22,17 +22,38 @@ As website changes require staff approval, discussing the page content beforehan
+│   ├── contributing.md
+│   └── bot.md
+└── _info.yml
+```
````
On one hand, it may be more convenient to keep all files related to a category within the same directory. On the other hand, it's less intuitive when considering how normal categories and pages are defined. Have you thought about moving the page out of the directory and instead detecting if the directory contains both a file and subdirectory with the same name (excluding the file extension)?

kosayoda replied:
I've thought of that. I personally find putting category pages in the category itself more intuitive.
If `contributing.md` is placed in the `guides` folder, what would be the description shown in the entry for `Contributing` in `..com/guides/`? The current implementation shows the information from `contributing/_info.yml`, while the rendered page has metadata from `contributing.md`.
Would placing `contributing.md` in the `guides` folder then make it confusing as to what information gets displayed where?

MarkKoz replied:
I see it as less intuitive because it breaks the established convention of a page being a file that is a direct child of the directory. Your point about info coming from different pages follows the same behaviour a normal page and category would follow. Really, the only deviation is that clicking the link in the parent category listing will lead to the page rather than to another category listing. It makes sense to me since I understand that when the parent lists it, it treats it as a normal category in every sense except for what is rendered when the link is clicked.

I don't feel strongly about this, but I think it's worth discussing at least.

MarkKoz replied:
Also note that with the current approach, it's impossible for it to contain a child page with the same name as the category. Granted, I don't foresee this being something we'd need to do.

kosayoda replied:
I see your point. The change won't be hard to make, the main addition is just to not list the category page in the listing if there is also a category with the same name in the directory.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -59,9 +79,9 @@ Pages, which include guides, articles, and other static content,...
-- **icon:** Favicon for the category entry for the page. Default: `fa-python` <i class="fab fa-python is-black" aria-hidden="true"></i>
+- **icon:** Icon for the category entry for the page. Default: `fab fa-python` <i class="fab fa-python is-black" aria-hidden="true"></i>
 - **relevant_links:** A YAML dictionary containing `text:link` pairs. See the example above.
+- **toc:** A number representing the smallest heading tag to show in the table of contents.
```
How to define a table of contents is currently not documented.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -22,17 +22,38 @@ As website changes require staff approval, discussing the page content beforehan
+
+In the above example, `www.pythondiscord.com/guides/` will list `Contributing` as a category entry with information from `contributing/_info.yml`.
+
+However, `www.pythondiscord.com/guides/contributing` will render `contributing.md` rather than show the category contents, so *it is the article's responsibility to link to any subpages under the article*.
```
Have you considered any way to show the direct children, or even all children? Perhaps a dropdown menu somewhere, like near the breadcrumbs.

kosayoda replied:
I have. Since I'm just trying to get feature parity with the current site, I'll open a separate issue for this so the app can move along.

kosayoda replied:
Plot twist: I did it anyways since it's not a big change. The dropdown only shows up on category-page-hybrids and look and function similar to the one shown on the current site when you're logged in.

The differences are that it only shows direct children, since any sub-categories show their children on their page anyway, and that the dropdown is click-toggled and not hover-toggled to be mobile-friendly.

kosayoda committed:
Place category pages in the same directory as categories.

kosayoda committed:
Document defining a table of contents for a page.

kosayoda committed:
Add a dropdown menu listing direct children.

This only shows when the page is also a category, since regular pages
have no children and regular categories already list their children.

kosayoda committed:
Allow clicking outside the dropdown to hide it.

kosayoda committed:
Update misleading documentation.

kosayoda committed:
Fix overriden PAGES_PATH in settings.py.

This was causing mysterious 404s because while migrating a feature to
dewikification, the same variable would be placed in the settings file.

MarkKoz commented on pydis_site/templates/content/dropdown.html:
```diff
@@ -0,0 +1,36 @@
+    <div class="dropdown-menu" id="subarticle-menu" role="menu">
+        <div class="dropdown-content">
+            {% for page in subarticles|dictsort:0 %}
+                <a href="{{page}}" class="dropdown-item">{{ page.title|replace_hyphens:" " }}</a>
```
While this is a simple and fast approach, it is too naïve and capitalises all words. Furthermore, file names may not necessarily be anything close to the titles in the metadata (unless we start saying that they should be).

kosayoda replied:
Hmm, I didn't think of that. I like the filename freedom, as we want the site URL to be as concise as possible, while having a more descriptive page title.

MarkKoz commented on pydis_site/apps/content/views/page_category.py:
```diff
@@ -33,29 +39,48 @@ def get_context_data(self, **kwargs) -> t.Dict[str, t.Any]:
+        if self.page_path.is_file() and self.category_path.is_dir():
+            context["subarticles"] = [
+                path.stem for path in self.category_path.iterdir()
+                if path.suffix != ".yml"
```
A strictly correct check would be `path.is_dir() or path.suffix == ".md"`.

kosayoda committed:
Use metadata titles in Sub-Articles dropdown.

This allows us to keep filenames (thus URLs) as concise as possible,
while having a more descriptive entry in the Sub-Articles dropdown for
category pages.
